### PR TITLE
May 2019 upgrade: add segwit recovery spec

### PR DESCRIPTION
**Pull request by floriansengel**
Specification of the exemption to the CLEANSTACK rule allowing the recovery of coins in segwit P2SH addresses.

### CODE REVIEW

**Comment by markblundeberg** [comment]
For clarity, add requirement that it applies to transaction inputs "that are pay-to-script-hash (P2SH)".

Just to make sure I am interpreting right:
* the first push has to be either OP_0 or OP_1 to OP_16; (OP_1NEGATE not allowed).
* the second push can use any of the four valid push forms (OP_PUSH<n>, OP_PUSHDATA1, OP_PUSHDATA2, OP_PUSHDATA4)?

**Comment by markblundeberg** [comment]
Hmm no, it looks like I misinterpreted -- only the *minimal* push form is allowed for the second push. The IsWitnessProgram() function imposes the following conditions:

* The redeemscript must be between 4 and 42 bytes in length.
* The first byte must be OP0 or OP_1 to OP_16.
* The second byte's value must be the length of the redeemscript minus two. (this will be between 2 and 40).

**Comment by floriansengel** [comment]
Fixed! It's more accurate now. I'm using the definition from [bip-141](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#witness-program)

File: spec/2019-05-15-segwit-recovery.md [hunk 1]
```diff
@@ -1,10 +1,27 @@
...
 In November 2018, the CLEANSTACK consensus rule was activated, with the intent of reducing malleability mechanisms. This had the unfortunate side effect of also making these segwit scripts *unspendable*, since attempting to spend these coins would always leave two items on the stack.
 
-Starting in May 2019, an exemption to the CLEANSTACK rule will be made:
+Starting in May 2019, an exemption to the CLEANSTACK rule will be made, allowing transactions spending segwit P2SH coins to be once again included in blocks.
 
-<Details to be provided>
+## Requirement
+A transaction input that:
+1. spends a P2SH coin; and
+1. the scriptSig contains solely the redeem script; and
```
Comment (markblundeberg) [inline spec/2019-05-15-segwit-recovery.md]: 2.

also, perhaps "the scriptSig execution only leaves the redeem script on stack" ?

File: spec/2019-05-15-segwit-recovery.md [hunk 2]
```diff
@@ -1,10 +1,27 @@
...
+2. the redeem script is a valid witness program: a 1-byte push opcode (for 0 to 16) followed by a data push between 2 and 40 bytes in minimal form;
```
Comment (markblundeberg) [inline spec/2019-05-15-segwit-recovery.md]: 3.

File: spec/2019-05-15-segwit-recovery.md [hunk 3]
```diff
@@ -1,10 +1,27 @@
...
+
+    shall be considered valid under the consensus rules to be activated in May 2019.
+
+## Examples
+
+Valid scriptSig recovering a P2SH-P2WPKH coin:
+
+    scriptSig='160014fcf9969ce1c98a135ed293719721fb69f0b686cb'
+ 
+Valid scripSig recovering a P2SH-P2WSH coin:
```
Comment (markblundeberg) [inline spec/2019-05-15-segwit-recovery.md]: scripSig -> scriptSig

File: spec/2019-05-15-segwit-recovery.md [hunk 4]
```diff
@@ -1,10 +1,27 @@
...
+Starting in May 2019, an exemption to the CLEANSTACK rule will be made, allowing transactions spending segwit P2SH coins to be once again included in blocks.
 
-<Details to be provided>
+## Requirement
+A transaction input that:
+1. spends a P2SH coin; and
+1. the scriptSig contains solely the redeem script; and
+2. the redeem script is a valid witness program: a 1-byte push opcode (for 0 to 16) followed by a data push between 2 and 40 bytes in minimal form;
+
+    shall be considered valid under the consensus rules to be activated in May 2019.
```
Comment (markblundeberg) [inline spec/2019-05-15-segwit-recovery.md]: deindent this; also, better to say "shall be exempted from the CLEANSTACK rule" as maybe the transaction input could be invalid due to some unrelated issue.

**Comment by Mengerian** [comment]
Great, looks good to me!

It makes sense to merge it now, and any needed improvements can be done after merging.

Thanks Florian and Mark for doing this!

**Pull request merged**